### ISSUE TITLE
Convert already-scheduled Central Office hearing to Virtual hearing

### ISSUE THREAD

**Issue by chandracarney** [department-of-veterans-affairs/caseflow#13804]
## Story
As a Hearing Coordinator, I want to convert a Veteran's already-scheduled Central Office hearing to a virtual hearing.

## Goals
Expand virtual hearings functionality so that Coordinators can convert any already-scheduled hearing _that's managed in Caseflow_ (i.e. Central Office or video hearing types) to a virtual hearing. This epic enables the conversion of already-scheduled Central Office hearings. A previous epic (#11508) enabled the conversion of already-scheduled video hearings.

## Success criteria
- From a user experience perspective, this design will be successful if Hearing Coordinators can easily convert a Central Office hearing to a virtual hearing, including selecting the Board Hearing Time without errors. We'll know the design is working if we observe that Coordinators understand how the Board Hearing Time relates to participants' timezones in usability tests (completed) and shadowing users in production (planned). 
- Quantitative metrics will focus on successful throughput and the ability of users to consistently convert Central Office hearings to virtual hearings.

## Out of scope
- Calculating video hearing times based on the Veteran's address, and the accompanying UI changes to the Daily Docket, will be addressed in later tickets (#12825 and potentially others)
- Displaying email previews
- Resending emails

## Release notes
All Caseflow hearing types can be converted to virtual using a full-page flow, allowing Hearing Coordinators to select the hearing time in the same step, and on the same page, as confirming Veteran and representative information.

More information, including address, is also displayed about the Veteran and representative. Because Veterans who are in the Central Office queue often opt to travel from outside of Washington, DC, the Veteran and representative timezones can be chosen so that emails are sent to all hearing parties in their timezone.

## Related tickets

#11158, #11508

## Designs
Figma: https://www.figma.com/file/V87TZArfdurCGJiEjQ73ES/Virtual-Hearings?node-id=203%3A14548

### Open questions
- [ ] Formatting of timezone names in dropdown fields for Veteran/Appellant and POA/Representative Timezone and virtual Hearing Time
- [ ] For some Service Organizations, a person's name, in addition to service org name, displays on the Daily Docket but not on Case Details. Is this name being pulled from VACOLS or VBMS? Should we try to display it here?
- [x] Custom Caseflow UI for the "you have unsaved changes" flow - can we use browser defaults for this iteration?
  - Yes, create another ticket for custom Caseflow UI
- [x] How much technical complexity would it add to have different layouts for the Virtual Hearing Details section depending on whether the hearing is formerly-Central or formerly-video? The h3 Veteran/POA headings aren't ideal for formerly-video now because we're not adding the timezone field yet.
  - We can have variations in display for formerly-Central and formerly-video hearings

### I. Updates to Hearing Details layout and style 
Addressed in `Add section cards, adjust layout and style of Hearing Details page` #14388

#### II.A Convert Central to virtual ([Figma link to start of flow](https://www.figma.com/file/V87TZArfdurCGJiEjQ73ES/Virtual-Hearings?node-id=5649%3A18611))

Addressed in `Build new Convert to Virtual page for Central Office hearings` (#14421)

#### II.B Update virtual (formerly-Central) hearing emails and/or timezones ([Figma link to start of flow](https://www.figma.com/file/V87TZArfdurCGJiEjQ73ES/Virtual-Hearings?node-id=5649%3A22954))
In general, when a user edits **either** the timezone **or** the email field on Hearing Details, a confirm modal subsequently displays **both** timezone **and** email. Specifically:
- When the veteran/appellant Timezone and/or email is edited (and is not blank since they're required fields), and the user presses the Save button, a modal displays ([link to design](https://www.figma.com/file/V87TZArfdurCGJiEjQ73ES/Virtual-Hearings?node-id=5649%3A23656)):
  - Veteran Email - read-only
  - Veteran Hearing Time - read-only
    - formatted as h:mm am/pm (hearing time in selected timezone) + timezone label
- When the POA/Representative Timezone and/or email is edited and the user presses the Save button, a modal displays ([link to design](https://www.figma.com/file/V87TZArfdurCGJiEjQ73ES/Virtual-Hearings?node-id=5649%3A24009)):
    - POA / Representative Email - read-only
    - POA / Representative Hearing Time - read-only
      - formatted as h:mm am/pm (hearing time in selected timezone) + timezone label
    - [Note that we don't currently handle deleting a previously-entered POA email - see issue #13945]
- When both the veteran/appellant and POA/Representative timezone and/or email are edited, a modal displaying info for both is displayed ([link to design](https://www.figma.com/file/V87TZArfdurCGJiEjQ73ES/Virtual-Hearings?node-id=5649%3A23286))
- On pressing the Update Virtual Hearing button, the async job to send emails begin, and the In Progress and and Success states/alerts display as appropriate
  - No changes to current state alert messaging (with variations for one or multiple participants) or form field disabled states, _except_ that the new Timezone fields are also disabled during the in progress state 

#### II.C Update virtual (formerly-Central) hearing time ([Figma link to start of flow](https://www.figma.com/file/V87TZArfdurCGJiEjQ73ES/Virtual-Hearings?node-id=5649%3A25353))
Addressed in `Update virtual (formerly-Central) hearing time` (#14640)

#### II.D Convert virtual hearing back to Central Office hearing ([Figma link to start of flow](https://www.figma.com/file/V87TZArfdurCGJiEjQ73ES/Virtual-Hearings?node-id=5649%3A20946))
Addressed in `Build new Convert to Virtual page for Central Office hearings` (#14421)

#### III.A Convert video to virtual ([Figma link to start of flow](https://www.figma.com/file/V87TZArfdurCGJiEjQ73ES/Virtual-Hearings?node-id=5649%3A26461))
Addressed in `Move to full-page for convert video to virtual flow` (#14642)

#### III.B Update virtual (formerly-video) hearing emails ([Figma link to start of flow](https://www.figma.com/file/V87TZArfdurCGJiEjQ73ES/Virtual-Hearings?node-id=5649%3A28721))
Only change to current state is to:
- update the modal's button styling to primary
- update the poa email field label from Representative Email to `POA / Representative Email`

#### III.C Update virtual (formerly-video) hearing time ([Figma link to start of flow](https://www.figma.com/file/V87TZArfdurCGJiEjQ73ES/Virtual-Hearings?node-id=5649%3A28743))
Only changes to current state are:
- change modal title to `Update Hearing Time` (for consistency with 'update' language used in similar situations elsewhere)
- change modal button's text to `Update Hearing Time` and its styling to primary

#### III.D Convert virtual hearing back to video ([Figma link](https://www.figma.com/file/V87TZArfdurCGJiEjQ73ES/Virtual-Hearings?node-id=5649%3A28010))
- Update "Convert to Video Hearing" functionality from the current modal to a full page, in order to maintain consistency with the new Central Office designs
- The Convert to Virtual Hearing page displays the following info / form fields:
  - Hearing Date - read-only
  - Hearing Time - same radio field that appears for Video hearings on the Daily Docket, except that the option values make the Regional Office timezone explicit, i.e. the option values format is [h:mm am/pm RegionalOfficeTimezone / h:mm am/pm BoardEasternTimezone]. Same formatting applies to values in the Other dropdown.
  - Regional Office - read-only
  - Hearing Location - same dropdown as on the Daily Docket
  - Veteran / Appellant name, mailing address, and email - read-only
  - Power of Attorney type label, POA name and/or service organization name, mailing address, email - read-only
     - if POA info is blank, display message ([direct link to design variation](https://www.figma.com/file/V87TZArfdurCGJiEjQ73ES/Virtual-Hearings?node-id=5649%3A21459))
- On pressing the Convert to Video Hearing button, the async job to send emails begin, and the In Progress and and Success states/alerts display as appropriate
  - No changes to current state alert messaging or form field disabled states

**Comment by ferristseng** [comment]
Blocked by an issue with figuring out what timezone to display the time in to the veteran.

**Comment by chandracarney** [comment]
Documenting on the story that we've gotten the go-ahead to [calculate using Veteran's timezone](https://dsva.slack.com/archives/C3EAF3Q15/p1585770263131600) from DVC McLeod to unblock us and move quickly on the Central Office use cases.

**Comment by carodew** [comment]
Would like to add some context – the copy to clipboard pattern that we've used elsewhere in Caseflow has had some accessibility issues that we've corrected on a one-off basis. Asking to make sure that the implementation in this case includes the fixes we've incorporated (I know @chandracarney will be able to provide guidance on this front).

**Comment by chandracarney** [comment]
Discussed this with the team and DVC McLeod. Since this use case has added complexity and needs design, we have decided to tackle several high-priority usability fixes in place of this added use case and will combine this use case in future work after the national rollout. Swapped this to another epic.
